### PR TITLE
Drop support for LLVM11

### PR DESCRIPTION
**Pull request by steven-johnson**
With Halide 13 released, we should drop support for LLVM11 in Halide trunk, since we only promise to support LLVM trunk + two releases.

### CODE REVIEW

**Comment by steven-johnson** [comment]
attn @alinas

**Comment by steven-johnson** [comment]
@alexreinking -- not sure about the packaging-for-Ubuntu workflow, I just updated to use LLVM12 instead, but not sure if that's kosher

File: src/CodeGen_LLVM.cpp [hunk 1]
```diff
@@ -1182,20 +1173,9 @@ void CodeGen_LLVM::optimize_module() {
                     asan_options, use_global_gc,
                     use_odr_indicator, destructor_kind));
             });
-#elif LLVM_VERSION >= 120
```
Comment (abadams) [inline src/CodeGen_LLVM.cpp:1185]: This looks like you deleted an llvm 12 block

File: test/correctness/simd_op_check.cpp [hunk 2]
```diff
@@ -1746,13 +1746,8 @@ class SimdOpCheck : public SimdOpCheckTest {
                     check("v128.const", 8 * w, f32_1 * f32(42));
                     check("v128.const", 4 * w, f64_1 * f64(42));
                 } else {
-                    if (Halide::Internal::get_llvm_version() == 120) {
```
Comment (abadams) [inline test/correctness/simd_op_check.cpp:1749]: This also looks like an llvm 12 block

**Review by abadams — APPROVED** [review]
lgtm except for two bits that confused me

File: src/CodeGen_LLVM.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/CodeGen_LLVM.cpp:1185]: No, I think it's just that the >=120 block was very similar to the else block here (just an extra arg to the lambda passed to registerPipelineStartEPCallback)

File: test/correctness/simd_op_check.cpp (hunk 2 above)
Comment (steven-johnson) [reply test/correctness/simd_op_check.cpp:1749]: it is, but we are keeping it

**Comment by alexreinking** [comment]
@steven-johnson - you need to update `11` to `12` here:

https://github.com/halide/Halide/blob/6071cf644b656b43a6de8b5d2d2d364d30678864/packaging/ubuntu/config.cmake#L63-L67

File: Makefile [hunk 3]
```diff
@@ -840,16 +840,12 @@ RUNTIME_LL_COMPONENTS = \
...
   x86_avx \
   x86_avx2 \
   x86_avx512 \
   x86_sse41
 
-ifeq (,$(findstring $(LLVM_VERSION_TIMES_10), 110 111))
-# x86_amx.ll won't compile under LLVM11, but we don't need it there, either
-RUNTIME_LL_COMPONENTS += x86_amx
-endif
-
```
Comment (alexreinking) [inline Makefile:852]: You might want to make a similar change here:

https://github.com/halide/Halide/blob/6071cf644b656b43a6de8b5d2d2d364d30678864/src/runtime/CMakeLists.txt#L116-L119

Comment (alexreinking) [reply Makefile:852]: This is the only usage of `Halide_LLVM_VERSION`, which is defined here and I am tempted to remove...

https://github.com/halide/Halide/blob/6071cf644b656b43a6de8b5d2d2d364d30678864/dependencies/llvm/CMakeLists.txt#L23-L24

Comment (steven-johnson) [reply Makefile:852]: > You might want to make a similar change here:
> 
> https://github.com/halide/Halide/blob/6071cf644b656b43a6de8b5d2d2d364d30678864/src/runtime/CMakeLists.txt#L116-L119

done

Comment (steven-johnson) [reply Makefile:852]: > This is the only usage of `Halide_LLVM_VERSION`, which is defined here and I am tempted to remove...
> 
> https://github.com/halide/Halide/blob/6071cf644b656b43a6de8b5d2d2d364d30678864/dependencies/llvm/CMakeLists.txt#L23-L24

I predict we will eventually need it again. Maybe keep with a comment.

File: test/correctness/simd_op_check.cpp (hunk 2 above)
Comment (abadams) [reply test/correctness/simd_op_check.cpp:1749]: But it's an == 120 block, not a >= 120 block, so by keeping it aren't we testing the wrong thing on llvm 13?

Or is the original check wrong?

Comment (abadams) [reply test/correctness/simd_op_check.cpp:1749]: bump. I'm confused about why this isn't going to cause simd op check failures on llvm 13+

Or were we only testing wasm on llvm 12 before?

**Comment by steven-johnson** [comment]
> I'm confused about why this isn't going to cause simd op check failures on llvm 13+

We were only testing on llvm14 before, actually, but I still think you are misreading the diff.

I'll test this locally on 12/13/14 to verify, though.

**Comment by steven-johnson** [comment]
> I'll test this locally on 12/13/14 to verify, though.

Passes for all three, so I think we are good.

File: test/correctness/simd_op_check.cpp (hunk 2 above)
Comment (abadams) [reply test/correctness/simd_op_check.cpp:1749]: Old code:
```
                    if (Halide::Internal::get_llvm_version() == 120) {
                        check("i64x2.splat", 8 * w, u16_1 * u16(42));
                        check("i64x2.splat", 4 * w, u32_1 * u32(42));
                    } else {
                        check("i16x8.splat", 8 * w, u16_1 * u16(42));
                        check("i32x4.splat", 4 * w, u32_1 * u32(42));
                   }
```

New code:

```
                    check("i64x2.splat", 8 * w, u16_1 * u16(42));
                    check("i64x2.splat", 4 * w, u32_1 * u32(42));
```

I think the new code is just the llvm_version() == 120 version, so aren't we changing the expected output for llvm 13/14?

I can't figure out what I'm missing.

**Comment by steven-johnson** [comment]
> I can't figure out what I'm missing.

You're missing the fact that the >= 130 case is handled above -- the section you're looking at is in the else case.

**Comment by abadams** [comment]
Ah! Everything is now clear.

**Pull request merged**